test: Fixes and updates to wallet_backwards_compatibility.py for 25.0 and descriptor wallets

achow101 opened the pull request:
It was somewhat surprising to me that wallet_backwards_compatibility.py did not catch #27915 since the purpose of the test is to find downgrade issues such as that. It turns out the test was deficient in several places when it came to testing descriptor wallets, as well as deficient in addition to failing to correctly test some releases.

This PR fixes these test cases, adds more informative logging, slightly refactors the entire test in order to better test future versions, and adds a 25.0 node to the test.

Notable changes:
* The compatibility test with 0.16 should not have been passing. The wallets were being copied incorrectly for 0.16 and resulting in 0.16 creating new wallets rather than testing the target wallets.
* The downgrade test will actually be run on descriptor wallets and it will test that downgrades are successful, and a subsequent upgrade is also successful. This catches #27915.
* The upgrade and downgrade test will be run on all versions up to master, rather than just 0.16, 0.17, and 0.19.

achow101 committed:
test: Run downgrade test on descriptor wallets

achow101 committed:
test: Run upgrade test on all nodes

achow101 committed:
test: Fix 0.16 wallet paths and downgrade test

The test for 0.16 wallet downgrading was using the wrong wallet path and
thus incorrectly finding that 0.16 could open wallets created in master.

achow101 committed:
test: Remove w1_v18 from wallet backwards compatibility

This wallet is no longer used in the test

achow101 committed:
test: Refactor v19 addmultisigaddress test to be distinct

This specific test is distinct from the rest of the backwards
compatibility tests as it is checking a specific failure.

achow101 committed:
test: add logging 0.17 incompatibilities in wallet back compat

achow101 committed:
test: Add 25.0 to wallet backwards compatibiilty test

achow101 committed:
test: Add 0.21 tr() incompatibility test

achow101 committed:
test: roundtrip wallet backwards compat downgrade

Test that old nodes don't mess up new wallets by loading a downgraded
wallet in master again.

Sjors commented:
Thanks for improving these tests. Added to review list.

ismaelsadeeq commented:
Concept ACK

S3RK commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -158,34 +173,46 @@ def run_test(self):
-                        continue
-                    node.loadwallet(wallet_name)
-                    wallet = node.get_wallet_rpc(wallet_name)
+                if node.version == 160300:
```
v0.16 is different in many regards, does it make sense to have a separate test to check incompatibility?

achow101 replied:
Perhaps, but I think we should test it in this test as it's important to have the upgrade and downgrade tests for the older wallet version.

S3RK replied:
agree, you can resolve this

achow101 replied:
No, this is more correct. This behavior doesn't apply to all older versions, only to 0.16.x. Versions prior to this kept the wallet files in the main datadir. 0.16 added the wallet directory but the wallets are still per file. 0.17 moved wallets into the current one directory per wallet structure.

achow101 commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -160,18 +155,25 @@ def run_test(self):
+                if node.version == 160300:
+                    # 0.16 node expect the wallet to be in the wallet dir but as a plain file rather than in directories
+                    shutil.copyfile(
+                        os.path.join(node_master.wallets_path, wallet, "wallet.dat"),
```
Done where the code was more than just moved.

achow101 commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -208,13 +235,29 @@ def run_test(self):
+        if self.options.descriptors:
+            self.log.info("Test descriptor wallet incompatibility on:")
+            for node in legacy_only_nodes:
+                if node.version < 180000:
```
Done

ismaelsadeeq reviewed (commented):
Running this test failed on my device not sure if it's from this PR though.

Steps to recreate
`git checkout pr/28027`
`make clean`
`./autogen.sh`
`./configure`
`make`
`test/functional/test_runner.py wallet_backwards_compatibility.py`
```
2023-07-11T00:04:00.468000Z TestFramework (INFO): PRNG seed is: 1124432296273551722
2023-07-11T00:04:00.468000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_banbrcm1
2023-07-11T00:04:00.472000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_framework.py", line 550, in start_nodes
    node.start(extra_args[i], *args, **kwargs)
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_node.py", line 221, in start
    self.process = subprocess.Popen(self.args + extra_args, env=subp_env, stdout=stdout, stderr=stderr, cwd=cwd, **kwargs)
  File "/usr/lib/python3.10/subprocess.py", line 969, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.10/subprocess.py", line 1845, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '/home/abubakarsadiq/Desktop/bitcoin/releases/v25.0/bin/bitcoind'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_framework.py", line 130, in main
    self.setup()
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_framework.py", line 295, in setup
    self.setup_network()
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_framework.py", line 389, in setup_network
    self.setup_nodes()
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/wallet_backwards_compatibility.py", line 74, in setup_nodes
    self.start_nodes()
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_framework.py", line 555, in start_nodes
    self.stop_nodes()
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_framework.py", line 570, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_node.py", line 346, in stop_node
    self.stop(wait=wait)
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_node.py", line 192, in __getattr__
    assert self.rpc_connected and self.rpc is not None, self._node_msg("Error: no RPC connection")
AssertionError: [node 0] Error: no RPC connection
2023-07-11T00:04:00.523000Z TestFramework (INFO): Stopping nodes
Traceback (most recent call last):
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/wallet_backwards_compatibility.py", line 368, in <module>
    BackwardsCompatibilityTest().main()
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_framework.py", line 154, in main
    exit_code = self.shutdown()
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_framework.py", line 311, in shutdown
    self.stop_nodes()
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_framework.py", line 570, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_node.py", line 346, in stop_node
    self.stop(wait=wait)
  File "/home/abubakarsadiq/Desktop/bitcoin/test/functional/test_framework/test_node.py", line 192, in __getattr__
    assert self.rpc_connected and self.rpc is not None, self._node_msg("Error: no RPC connection")
AssertionError: [node 0] Error: no RPC connection
[node 1] Cleaning up leftover process
[node 0] Cleaning up leftover process
```

Left some comments and question

S3RK commented:
Approach ACK

achow101 commented:
> Running this test failed on my device not sure if it's from this PR though.

It looks like you're missing the 25.0 binaries. The test is expecting to find 25.0's bitcoind at `/home/abubakarsadiq/Desktop/bitcoin/releases/v25.0/bin/bitcoind` but doesn't.

achow101 commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -72,10 +72,45 @@ def setup_nodes(self):
+        assert info['private_keys_enabled']
+        assert info['keypoolsize'] > 0
+        # Use addmultisigaddress (see #18075)
+        address_18075 = wallet.rpc.addmultisigaddress(1, ["0296b538e853519c726a2c91e61ec11600ae1390813a627c66fb8be7947be63c52", "037211a824f55b505228e4c3d5194c1fcfaa15a456abdf37f9b9d97a4040afc073"], "", "legacy")["address"]
```
I don't think the specific keys are relevant.

achow101 commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -127,13 +126,6 @@ def run_test(self):
         assert wallet.getaddressinfo(address_18075)["solvable"]
         node_v19.unloadwallet("w1_v19")
 
-        # w1_v18: regular wallet, created with v0.18
```
I would guess that it was used in the up-downgrade test but was accidentally removed at some point. It isn't necessary anymore either as that test is now automated to go over all of the previous versions so the hardcoded specific wallet versions are no longer necessary.

achow101 commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -192,52 +194,60 @@ def run_test(self):
+                continue
+            self.log.info(f"- {node.version}")
+            for wallet_name in ["w1", "w2", "w3"]:
+                if node.version < 180000 and wallet_name == "w3":
```
Will do if I need to retouch.

achow101 replied:
Instead of using magic numbers, I've changed the test to use helper functions that check the major version.

Sjors commented:
0ffa0f1c370494b6b99a9cdf911b20ee655ac829 (and later) breaks at line line 96 when you run `wallet_backwards_compatibility.py --descriptors` without BDB:

```
JSONRPCException: Wallet file verification failed. Failed to open database path '/tmp/bitcoin_func_test_x3n9_0vn/node1/regtest/wallets/w1_v19'. Build does not support Berkeley DB database format. (-18)
```

Otherwise it works, just some stylistic notes inline.

> 0.16 creating new wallets rather than testing the target wallets

Is there an assert you can add in c31e0db7f4263096d503f07707af9d0a733e50da that would catch this directly? Right now a regression would be "caught" simply because it doesn't fail to load. Ideally we'd check this for every version before we dropped the "just create a new one if it doesn't exist" behavior was dropped.

achow101 commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -76,7 +76,6 @@ def run_test(self):
         node_miner = self.nodes[0]
         node_master = self.nodes[1]
         node_v19 = self.nodes[self.num_nodes - 4]
```
0.18 is still tested, just not explicitly with its own special casing and variables.

achow101 committed:
test: Add helper functions for checking node versions

achow101 commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -261,6 +261,12 @@ def run_test(self):
+        # When descriptors are enabled, w1 cannot be opened by 0.21 sinced it contains a taproot descriptor
+        if self.options.descriptors:
+            self.log.info("Test that 0.21 cannot open wallet containing tr() descriptors")
+            node_v21 = self.nodes[-6]
```
Done

achow101 commented:
> [0ffa0f1](https://github.com/bitcoin/bitcoin/commit/0ffa0f1c370494b6b99a9cdf911b20ee655ac829) (and later) breaks at line line 96 when you run `wallet_backwards_compatibility.py --descriptors` without BDB:

Fixed

> > 0.16 creating new wallets rather than testing the target wallets
> 
> Is there an assert you can add in [c31e0db](https://github.com/bitcoin/bitcoin/commit/c31e0db7f4263096d503f07707af9d0a733e50da) that would catch this directly? Right now a regression would be "caught" simply because it doesn't fail to load. Ideally we'd check this for every version before we dropped the "just create a new one if it doesn't exist" behavior was dropped.

With the versions that have `loadwallet`, we catch that case by checking the contents of the wallet. The test previously wasn't doing that for 0.16 (it was actually one of the first things I changed before I realized there was an incompatibility).

furszy commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -178,18 +178,25 @@ def run_test(self):
             for wallet in os.listdir(node_master_wallets_dir):
-                shutil.copytree(
-                    os.path.join(node_master_wallets_dir, wallet),
-                    os.path.join(self.nodes_wallet_dir(node), wallet)
```
In df850c25:
Removed the only usage of the `nodes_wallet_dir` function but not the, now unused, function.

achow101 replied:
Done

furszy commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -178,18 +178,25 @@ def run_test(self):
+                    shutil.copytree(
+                        node_master_wallets_dir / wallet,
+                        node.wallets_path / wallet,
+                    )
```
In df850c25:

Could simplify this:

```suggestion
            # Copy wallets to previous version
            for wallet in os.listdir(node_master_wallets_dir):
                source = node_master_wallets_dir / wallet
                if self.major_version_equals(node, 16):
                    # 0.16 node expect the wallet to be in the wallet dir but as a plain file rather than in directories
                    source / "wallet.dat"
                shutil.copytree(source, node.wallets_path / wallet)
```

achow101 replied:
Done

furszy commented on test/functional/wallet_backwards_compatibility.py:
```diff
@@ -208,130 +261,133 @@ def run_test(self):
+            else:
+                node_v16.assert_start_raises_init_error([f"-wallet={wallet_name}"], f"Error: Error loading {wallet_name}: Wallet requires newer version of Bitcoin Core")
+
+        # When descriptors are enabled, w1 cannot be opened by 0.21 sinced it contains a taproot descriptor
```
s/sinced/since

achow101 replied:
Done

Sjors commented:
ACK 91f47e676ae3f131e7cecae9b5b5e50b358e9b32

Also tested that reverting 6a9510d2dabde1c9ee6c4226e3d16ca32eb48ac5 causes the test to fail when it tries to load the wallet in master after having loaded it in v25 (`Wallet corrupted`).

furszy reviewed (commented):
ACK afd9a67

Sjors commented:
re-ACK afd9a673c458e97305da49a70a1ddbf60e651876
